File: bridges/usb-serial-emg-receiver.py
#!/usr/bin/env python3
"""
USB Serial EMG Receiver (Python)

Reads EMG JSON data from an ESP32 via USB serial and POSTs it to the
Next.js API at /api/emg/ws. Same function as usb-serial-emg-receiver.js
but uses pyserial instead of the Node.js serialport module.

Usage:
  python bridges/usb-serial-emg-receiver.py              # auto-detect ESP32
  python bridges/usb-serial-emg-receiver.py COM4          # specify port
  python bridges/usb-serial-emg-receiver.py COM4 115200   # specify port + baud

Env:
  EMG_SERIAL_PORT=auto|COMx   EMG_BAUD_RATE=115200
  EMG_API_URL=http://127.0.0.1:3000/api/emg/ws
  DEBUG_EMG=1

Requirements: pip install pyserial requests
"""

import json
import os
import sys
import time
import threading

try:
    import serial
    import serial.tools.list_ports
except ImportError:
    print("pyserial is required: pip install pyserial")
    sys.exit(1)

try:
    import requests
except ImportError:
    print("requests is required: pip install requests")
    sys.exit(1)

TAG = "[emg-usb-py]"

API_URL = os.environ.get("EMG_API_URL", "http://127.0.0.1:3000/api/emg/ws")
SERIAL_PORT = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("EMG_SERIAL_PORT", "")
BAUD = int(sys.argv[2]) if len(sys.argv) > 2 else int(os.environ.get("EMG_BAUD_RATE", "115200"))
DEBUG = os.environ.get("DEBUG_EMG", "").strip().lower() in ("1", "true", "yes")
RECONNECT_DELAY = 5

PI_VID_PID = ("0525", "A4A7")
ESP32_KEYWORDS = ["cp210", "ch340", "ch341", "ftdi", "silicon", "esp", "wch"]

stats = {"posted": 0, "failures": 0, "parse_errors": 0, "skipped": 0}


def is_esp32_port(p):
    desc = ((p.description or "") + " " + (p.manufacturer or "")).lower()
    if p.vid and f"{p.vid:04X}" != PI_VID_PID[0]:
        for kw in ESP32_KEYWORDS:
            if kw in desc:
                return True
        if p.vid:
            return True
    return False


def is_bluetooth(p):
    desc = ((p.description or "") + " " + (p.manufacturer or "") + " " + (p.hwid or "")).lower()
    return "bluetooth" in desc or "bthenum" in desc


def is_raspberry_pi(p):
    return p.vid and f"{p.vid:04X}" == PI_VID_PID[0] and f"{p.pid:04X}" == PI_VID_PID[1]


def auto_detect():
    print(f"{TAG} scanning for ESP32/MCU ports...\n")
    ports = sorted(serial.tools.list_ports.comports(), key=lambda x: x.device)

    for p in ports:
        vid_pid = f"{p.vid:04X}:{p.pid:04X}" if p.vid else "-"
        print(f"  {p.device:<10} {p.description or '?':<40} {vid_pid}")
    print()

    candidates = [p for p in ports if is_esp32_port(p) and not is_bluetooth(p) and not is_raspberry_pi(p)]

    if not candidates:
        non_bt = [p for p in ports if not is_bluetooth(p) and not is_raspberry_pi(p)]
        candidates = non_bt

    for p in candidates:
        print(f"{TAG} probing {p.device} @ {BAUD}...")
        try:
            ser = serial.Serial(p.device, BAUD, timeout=6)
            deadline = time.time() + 6
            while time.time() < deadline:
                if ser.in_waiting:
                    line = ser.readline().decode("utf-8", errors="replace").strip()
                    if looks_like_emg(line):
                        ser.close()
                        print(f"{TAG} EMG data detected on {p.device}\n")
                        return p.device
                time.sleep(0.05)
            ser.close()
            print(f"{TAG}   no EMG data on {p.device}")
        except serial.SerialException as e:
            print(f"{TAG}   cannot open {p.device}: {e}")

    print(f"{TAG} no ESP32/MCU with EMG data found.")
    return None


def looks_like_emg(line):
    try:
        obj = json.loads(line)
        return (
            obj.get("type") in ("emg_data", "calibration_data", "heartbeat", "test")
            or isinstance(obj.get("muscleActivity"), (int, float))
            or isinstance(obj.get("voltage"), (int, float))
        )
    except (json.JSONDecodeError, ValueError):
        return False


def post_data(payload):
    try:
        r = requests.post(API_URL, json=payload, timeout=5)
        if r.ok:
            stats["posted"] += 1
        else:
            stats["failures"] += 1
            if DEBUG:
                print(f"{TAG} API {r.status_code}")
    except Exception as e:
        stats["failures"] += 1
        if stats["failures"] <= 3 or stats["failures"] % 50 == 0:
            print(f"{TAG} API error: {e} (failures: {stats['failures']})")


def log_summary():
    while True:
        time.sleep(10)
        if stats["posted"] > 0 or stats["failures"] > 0:
            print(f"{TAG} summary: posted={stats['posted']} failures={stats['failures']} parse_errors={stats['parse_errors']} skipped={stats['skipped']}")


def run(port_path):
    print(f"{TAG} opening {port_path} @ {BAUD}...")
    ser = serial.Serial(port_path, BAUD, timeout=1)
    print(f"{TAG} connected — forwarding to {API_URL}")
    print(f"{TAG} debug: {'on' if DEBUG else 'off'}\n")

    first_frame = True

    try:
        while True:
            if ser.in_waiting:
                line = ser.readline().decode("utf-8", errors="replace").strip()
                if not line or not line.startswith("{"):
                    stats["skipped"] += 1
                    continue

                try:
                    obj = json.loads(line)
                except json.JSONDecodeError:
                    stats["parse_errors"] += 1
                    continue

                if "type" not in obj:
                    if isinstance(obj.get("muscleActivity"), (int, float)):
                        obj["type"] = "emg_data"
                    else:
                        stats["skipped"] += 1
                        continue

                if first_frame:
                    first_frame = False
                    print(f"{TAG} first frame: type={obj.get('type')} muscleActivity={obj.get('muscleActivity')} voltage={obj.get('voltage')}")

                post_data(obj)
            else:
                time.sleep(0.01)
    except serial.SerialException as e:
        print(f"{TAG} serial error: {e}")
    except KeyboardInterrupt:
        print(f"\n{TAG} stopped.")
    finally:
        ser.close()


def main():
    port = SERIAL_PORT.strip()
    if not port or port.lower() == "auto":
        port = auto_detect()
        if not port:
            print(f"\n{TAG} Could not find an ESP32/MCU.")
            print(f"{TAG} Plug in the ESP32 via USB and try again, or specify: python {sys.argv[0]} COM4")
            sys.exit(1)

    summary_thread = threading.Thread(target=log_summary, daemon=True)
    summary_thread.start()

    while True:
        try:
            run(port)
        except serial.SerialException as e:
            print(f"{TAG} {e} — reconnecting in {RECONNECT_DELAY}s...")
        except KeyboardInterrupt:
            break
        time.sleep(RECONNECT_DELAY)


if __name__ == "__main__":
    main()

File: bridges/usb-serial-emg-receiver.js
/**
 * USB Serial EMG Receiver — MyoWare/ESP32 via USB CDC → POST /api/emg/ws → CCA UI
 *
 * Reads EMG data from an ESP32 (or similar MCU) connected via USB serial,
 * parses JSON lines, and forwards them to the Next.js EMG API endpoint.
 *
 * Data format expected from ESP32 (one JSON object per line):
 *   {"type":"emg_data","timestamp":12345,"muscleActivity":1024,"voltage":0.82,"calibrated":false}
 *   {"type":"calibration_data","min":100,"max":3500,"range":3400,"timestamp":12345}
 *   {"type":"heartbeat","timestamp":12345}
 *
 * Env (.env / .env.local):
 *   EMG_SERIAL_PORT=auto | COMx         (auto = scan for ESP32/MCU USB, default auto)
 *   EMG_BAUD_RATE=115200                (default 115200)
 *   EMG_API_URL=http://127.0.0.1:3000/api/emg/ws
 *   DEBUG_EMG=1                         (verbose logging)
 *   EMG_PROBE_MS=6000                   (auto-detect timeout per port)
 *   EMG_RECONNECT_MS=5000              (delay before reconnecting on disconnect)
 */

const path = require("path");
require("dotenv").config({ path: path.join(__dirname, "..", ".env") });
require("dotenv").config({ path: path.join(__dirname, "..", ".env.local") });

const { SerialPort } = require("serialport");
const { ReadlineParser } = require("@serialport/parser-readline");

const TAG = "[emg-usb]";

function isTruthyEnv(v) {
  if (v === undefined || v === null) return false;
  const s = String(v).trim().toLowerCase();
  return s === "1" || s === "true" || s === "yes";
}

function resolveApiUrl() {
  const raw = process.env.EMG_API_URL || "http://127.0.0.1:3000/api/emg/ws";
  return raw.includes("localhost") ? raw.replace(/localhost/g, "127.0.0.1") : raw;
}

const API_URL = resolveApiUrl();

let SERIAL_PORT = process.argv[2] || process.env.EMG_SERIAL_PORT || "";
{
  const sp = String(SERIAL_PORT).trim().toLowerCase();
  if (sp === "" || sp === "auto") SERIAL_PORT = "";
}

const DEFAULT_BAUD =
  parseInt(process.argv[3] || process.env.EMG_BAUD_RATE || "115200", 10) || 115200;

const DEBUG = isTruthyEnv(process.env.DEBUG_EMG);
const PROBE_MS = Math.max(1000, parseInt(process.env.EMG_PROBE_MS || "6000", 10) || 6000);
const RECONNECT_MS = parseInt(process.env.EMG_RECONNECT_MS || "5000", 10) || 5000;

const stats = { posted: 0, failures: 0, parseErrors: 0, skipped: 0 };
let logSummaryInterval = null;

// Known ESP32/MCU USB identifiers
const MCU_PATTERNS = [
  /cp210/i, /ch340/i, /ch341/i, /ftdi/i, /silicon.lab/i,
  /esp/i, /wch/i, /arduino/i, /teensy/i,
];

function isMcuPort(port) {
  const desc = [
    port.manufacturer || "",
    port.pnpId || "",
    port.friendlyName || "",
  ].join(" ");
  if (port.vendorId && !/^0*$/.test(port.vendorId)) return true;
  return MCU_PATTERNS.some((re) => re.test(desc));
}

function isBluetoothPort(port) {
  const desc = [
    port.manufacturer || "",
    port.pnpId || "",
    port.friendlyName || "",
  ].join(" ").toLowerCase();
  return desc.includes("bluetooth") || desc.includes("bthenum");
}

function isRaspberryPiPort(port) {
  return (
    (port.vendorId === "0525" && port.productId === "A4A7") ||
    /raspberry/i.test(port.friendlyName || "")
  );
}

function looksLikeEmgPayload(line) {
  try {
    const obj = JSON.parse(line.trim());
    return (
      obj.type === "emg_data" ||
      obj.type === "calibration_data" ||
      obj.type === "heartbeat" ||
      obj.type === "test" ||
      typeof obj.muscleActivity === "number" ||
      typeof obj.voltage === "number"
    );
  } catch {
    return false;
  }
}

async function sortPortsMcuFirst(ports) {
  const mcu = [];
  const other = [];
  for (const p of ports) {
    if (isBluetoothPort(p) || isRaspberryPiPort(p)) continue;
    if (isMcuPort(p)) mcu.push(p);
    else other.push(p);
  }
  return [...mcu, ...other];
}

function probePort(portPath, baud, timeoutMs) {
  return new Promise((resolve) => {
    let found = false;
    const timer = setTimeout(() => {
      if (!found) {
        found = true;
        try { sp.close(); } catch {}
        resolve(false);
      }
    }, timeoutMs);

    let sp;
    try {
      sp = new SerialPort({ path: portPath, baudRate: baud, autoOpen: false });
    } catch {
      clearTimeout(timer);
      resolve(false);
      return;
    }

    const parser = sp.pipe(new ReadlineParser({ delimiter: "\n" }));

    parser.on("data", (line) => {
      if (!found && looksLikeEmgPayload(line)) {
        found = true;
        clearTimeout(timer);
        try { sp.close(); } catch {}
        resolve(true);
      }
    });

    sp.on("error", () => {
      if (!found) {
        found = true;
        clearTimeout(timer);
        resolve(false);
      }
    });

    sp.open((err) => {
      if (err) {
        if (!found) {
          found = true;
          clearTimeout(timer);
          resolve(false);
        }
      }
    });
  });
}

async function autoDetectPort() {
  console.log(`${TAG} auto-detect: scanning for ESP32/MCU with EMG data...\n`);

  const allPorts = await SerialPort.list();
  console.log(`${TAG} detected serial ports (${allPorts.length}):`);
  for (const p of allPorts) {
    console.log(`   ${p.path} - ${p.friendlyName || p.manufacturer || "unknown"} ${p.vendorId ? `(VID:${p.vendorId})` : ""}`);
  }

  const sorted = await sortPortsMcuFirst(allPorts);
  if (sorted.length === 0) {
    console.error(`${TAG} no non-Bluetooth, non-Pi serial ports found.`);
    return null;
  }

  console.log(`${TAG} probe order (MCU-like first, Bluetooth/Pi excluded):`);
  sorted.forEach((p) => console.log(`   ${p.path}`));
  console.log();

  for (const portInfo of sorted) {
    console.log(`${TAG} probing ${portInfo.path} @ ${DEFAULT_BAUD} (${PROBE_MS}ms)...`);
    const ok = await probePort(portInfo.path, DEFAULT_BAUD, PROBE_MS);
    if (ok) {
      console.log(`${TAG} EMG data detected on ${portInfo.path} @ ${DEFAULT_BAUD}\n`);
      return portInfo.path;
    }
    console.log(`${TAG}    no EMG data on ${portInfo.path}`);
  }

  console.error(`\n${TAG} no EMG stream found on any port.`);
  return null;
}

async function forwardToApi(payload) {
  try {
    const res = await fetch(API_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (res.ok) {
      stats.posted++;
    } else {
      stats.failures++;
      if (DEBUG) console.error(`${TAG} API ${res.status}: ${res.statusText}`);
    }
  } catch (err) {
    stats.failures++;
    if (stats.failures <= 3 || stats.failures % 50 === 0) {
      console.error(`${TAG} API error: ${err.message} (failures: ${stats.failures})`);
    }
  }
}

function logSummary() {
  const { posted, failures, parseErrors, skipped } = stats;
  if (posted > 0 || failures > 0) {
    console.log(`${TAG} forward summary { posted: ${posted}, failures: ${failures}, parseErrors: ${parseErrors}, skipped: ${skipped} }`);
  }
}

async function connectAndStream(portPath, baud) {
  console.log(`${TAG} opening ${portPath} @ ${baud}...`);

  const sp = new SerialPort({ path: portPath, baudRate: baud, autoOpen: false });
  const parser = sp.pipe(new ReadlineParser({ delimiter: "\n" }));

  let firstFrame = true;

  parser.on("data", (line) => {
    const trimmed = line.trim();
    if (!trimmed || !trimmed.startsWith("{")) {
      stats.skipped++;
      return;
    }

    let obj;
    try {
      obj = JSON.parse(trimmed);
    } catch {
      stats.parseErrors++;
      if (DEBUG) console.warn(`${TAG} parse error: ${trimmed.substring(0, 100)}`);
      return;
    }

    if (!obj.type) {
      if (typeof obj.muscleActivity === "number" || typeof obj.voltage === "number") {
        obj.type = "emg_data";
      } else {
        stats.skipped++;
        return;
      }
    }

    if (firstFrame) {
      firstFrame = false;
      console.log(`${TAG} first EMG frame received:`, {
        type: obj.type,
        muscleActivity: obj.muscleActivity,
        voltage: obj.voltage,
        calibrated: obj.calibrated,
      });
    }

    forwardToApi(obj);
  });

  sp.on("close", () => {
    console.warn(`${TAG} serial port closed. Reconnecting in ${RECONNECT_MS}ms...`);
    clearInterval(logSummaryInterval);
    setTimeout(() => main(), RECONNECT_MS);
  });

  sp.on("error", (err) => {
    console.error(`${TAG} serial error: ${err.message}`);
  });

  return new Promise((resolve, reject) => {
    sp.open((err) => {
      if (err) {
        reject(err);
        return;
      }
      console.log(`${TAG} connected to ${portPath} @ ${baud}`);
      console.log(`${TAG} forwarding EMG data → ${API_URL}`);
      console.log(`${TAG} debug: ${DEBUG ? "on" : "off"}\n`);

      logSummaryInterval = setInterval(logSummary, 10000);
      resolve();
    });
  });
}

async function waitForNextJs(url, maxRetries = 10) {
  const healthUrl = url.replace(/\/api\/emg\/ws$/, "");
  for (let i = 0; i < maxRetries; i++) {
    try {
      const res = await fetch(healthUrl, { method: "HEAD" });
      if (res.ok || res.status === 404 || res.status === 405) {
        console.log(`${TAG} Next.js reachable at ${healthUrl}`);
        return true;
      }
    } catch {}
    if (i === 0) console.log(`${TAG} waiting for Next.js at ${healthUrl}...`);
    await new Promise((r) => setTimeout(r, 2000));
  }
  console.warn(`${TAG} Next.js not reachable — will forward anyway (may fail initially)`);
  return false;
}

async function main() {
  let portPath = SERIAL_PORT;

  if (!portPath) {
    portPath = await autoDetectPort();
    if (!portPath) {
      console.error(`\n${TAG} Could not find an ESP32/MCU with EMG data.`);
      console.error(`${TAG} Troubleshooting:`);
      console.error(`${TAG}   1. Make sure the ESP32 is plugged in via USB`);
      console.error(`${TAG}   2. Check Device Manager → Ports for the COM port`);
      console.error(`${TAG}   3. Make sure the ESP32 firmware is running and sending data`);
      console.error(`${TAG}   4. Try specifying the port: node bridges/usb-serial-emg-receiver.js COM4`);
      console.error(`${TAG}   5. Set EMG_SERIAL_PORT=COMx in .env.local`);
      process.exit(1);
    }
  }

  await waitForNextJs(API_URL);

  try {
    await connectAndStream(portPath, DEFAULT_BAUD);
  } catch (err) {
    console.error(`${TAG} failed to open ${portPath}: ${err.message}`);
    console.error(`${TAG} retrying in ${RECONNECT_MS}ms...`);
    setTimeout(() => main(), RECONNECT_MS);
  }
}

main();
